src/ucp/dt/datatype_iter.inl

```diff
@@ -502,10 +550,17 @@ ucp_datatype_iter_is_end(const ucp_datatype_iter_t *dt_iter)
 static UCS_F_ALWAYS_INLINE void
 ucp_datatype_memh_dereg(ucp_context_h context, ucp_mem_h *memh_p)
 {
-    if (*memh_p != NULL) {
-        ucp_memh_put(context, *memh_p);
-        *memh_p = NULL;
+    if (*memh_p == NULL) {
+        return;
     }
+
+    /* Do no release user memory handle */
+    if (ucp_memh_is_user_memh(*memh_p)) {
+        return;
+    }
+
+    ucp_memh_put(context, *memh_p);
+    *memh_p = NULL;
```

Comment:
if it's a user memh, IMO still need to set *memh_p = NULL

Reply:
can you explain why?
Isn't it the user responsibility? (by calling mem_unmap?)

Reply:
need to remove reference to this memh from ucp_request_t so next time when it will be reused if in line 607 will work as expected

Reply:
seems it does not work with RNDV internal requests, which call reg/dereg multiple times without going threw init.
The correct place to set memh to NULL is in cleanup.

Reply:
```
static UCS_F_ALWAYS_INLINE void
ucp_datatype_iter_cleanup(ucp_datatype_iter_t *dt_iter, unsigned dt_mask)
{
    if (ucp_datatype_iter_is_class(dt_iter, UCP_DATATYPE_CONTIG, dt_mask)) {
        dt_iter->type.contig.memh = NULL;
    } else if (ucp_datatype_iter_is_class(dt_iter, UCP_DATATYPE_IOV, dt_mask)) {
        ucs_free(dt_iter->type.iov.memh);
        dt_iter->type.iov.memh = NULL;
    } else if (ucp_datatype_iter_is_class(dt_iter, UCP_DATATYPE_GENERIC,
                                          dt_mask)) {
        dt_iter->type.generic.dt_gen->ops.finish(dt_iter->type.generic.state);
    }
}
```

Reply:
ok